pkg/datapath/linux/devices_controller.go

```diff
@@ -0,0 +1,690 @@
+	}
+	for _, route := range routes {
+		batch[route.LinkIndex] = append(batch[route.LinkIndex], netlink.RouteUpdate{
+			Type:  unix.RTM_NEWROUTE,
+			Route: route,
+		})
+	}
+
+	if !dc.processBatch(true, batch) {
+		// This can only happen if the statedb rejects the commit.
+		return errors.New("internal error: initial processing of devices failed")
+	}
+
+	select {
+	case <-dc.initialized:
+	default:
+		close(dc.initialized)
+	}
+
+	return nil
+}
+
+func (dc *devicesController) processUpdates(
+	ctx context.Context,
+	addrUpdates chan netlink.AddrUpdate,
+	routeUpdates chan netlink.RouteUpdate,
+	linkUpdates chan netlink.LinkUpdate,
+) {
+	// Collect all updates into a batch to minimize the
+	// number of write transactions when there are large
+	// amount of updates.
+	timer := time.NewTimer(batchingDuration)
+	// Start with a stopped timer and only start it when an
+	// update is appended.
+	if !timer.Stop() {
+		<-timer.C
+	}
+	timerStopped := true
+	defer func() {
+		if !timerStopped && !timer.Stop() {
+			<-timer.C
+		}
+	}()
+
+	batch := map[int][]any{}
+	appendUpdate := func(index int, u any) {
+		batch[index] = append(batch[index], u)
+		if timerStopped {
+			timer.Reset(batchingDuration)
+			timerStopped = false
+		}
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		case u, ok := <-addrUpdates:
+			if !ok {
+				return
+			}
+			appendUpdate(u.LinkIndex, u)
+
+		case r, ok := <-routeUpdates:
+			if !ok {
+				return
+			}
+			appendUpdate(r.LinkIndex, r)
+
+		case l, ok := <-linkUpdates:
+			if !ok {
+				return
+			}
+			appendUpdate(int(l.Index), l)
+
+		case <-timer.C:
+			timerStopped = true
+
+			if dc.processBatch(false, batch) {
+				batch = map[int][]any{}
+			}
+		}
+	}
+}
+
+func deviceAddressFromAddrUpdate(upd netlink.AddrUpdate) tables.DeviceAddress {
+	return tables.DeviceAddress{
+		Addr:  ip.MustAddrFromIP(upd.LinkAddress.IP),
+		Flags: upd.Flags,
+		Scope: upd.Scope,
+	}
+}
+
+func interfaceFromLink(link netlink.Link) net.Interface {
+	a := link.Attrs()
+	return net.Interface{
+		Index:        a.Index,
+		MTU:          a.MTU,
+		Name:         a.Name,
+		HardwareAddr: a.HardwareAddr,
+		Flags:        a.Flags,
+	}
+}
+
+// processBatch processes a batch of link, route and address updates. If it successfully
+// commits it to the state it returns true, if the batch is incomplete (e.g. route updates
+// for non-existing link), it returns false and the batch should be expanded with newer
+// updates and tried again.
+func (dc *devicesController) processBatch(init bool, batch map[int][]any) (ok bool) {
+	txn := dc.params.DB.WriteTxn()
+	devicesWriter := dc.params.DeviceTable.Writer(txn)
+	routesWriter := dc.params.RouteTable.Writer(txn)
+
+	if init {
+		// Initial commit, flush any leftover state from potentially
+		// previous failed processing loop. We're flushing it here
+		// in the same write transaction as the updates so that observers
+		// do not see the tables emptying.
+		dc.flushState(txn)
```

Comment:
Yeah, could log the error in `flushState`. Or might even be better to `Fatal` since failure in `DeleteAll` means we've messed up the query/indexes. Will fix this.